internal/datastore/postgres/snapshot.go

```diff
@@ -0,0 +1,236 @@
+}
+
+type comparisonResult uint8
+
+const (
+	uncomparable comparisonResult = iota
+	equal
+	lt
+	gt
+	concurrent
+)
+
+func (s pgSnapshot) compare(rhs pgSnapshot) comparisonResult {
+	if s.status != pgtype.Present || rhs.status != pgtype.Present {
+		return uncomparable
+	}
+
+	var rhsHasMoreInfo bool
+	for _, txid := range append(s.xipList, s.xmax) {
+		if rhs.visible(txid) {
+			rhsHasMoreInfo = true
+			break
+		}
+	}
+
+	var lhsHasMoreInfo bool
+	for _, txid := range append(rhs.xipList, rhs.xmax) {
+		if s.visible(txid) {
+			lhsHasMoreInfo = true
+			break
+		}
+	}
+
+	switch {
+	case rhsHasMoreInfo && lhsHasMoreInfo:
+		return concurrent
+	case rhsHasMoreInfo:
+		return lt
+	case lhsHasMoreInfo:
+		return gt
+	default:
+		return equal
+	}
+}
+
+func (s pgSnapshot) markComplete(txid uint64) pgSnapshot {
+	if txid < s.xmin || s.status != pgtype.Present {
+		// Nothing to do
+		return s
+	}
+
+	newSnapshot := pgSnapshot{
+		s.xmin,
+		s.xmax,
+		s.xipList,
+		pgtype.Present,
+	}
+
+	// Adjust the xmax and running tx if necessary
+	if txid >= s.xmax {
+		for newIP := s.xmax; newIP < txid+1; newIP++ {
+			newSnapshot.xipList = append(newSnapshot.xipList, newIP)
+		}
+		newSnapshot.xmax = txid + 1
+	}
+
+	// Mark the tx complete if it's in the xipList
+	// Note: we only find the first if it was erroneously duplicate
+	pos, found := slices.BinarySearch(newSnapshot.xipList, txid)
+	if found {
+		newSnapshot.xipList = slices.Delete(newSnapshot.xipList, pos, pos+1)
+	}
+
+	// Adjust the xmin if necessary
+	if len(newSnapshot.xipList) > 0 {
+		newSnapshot.xmin = newSnapshot.xipList[0]
+	} else {
+		newSnapshot.xmin = newSnapshot.xmax
+		newSnapshot.xipList = nil
+	}
+
+	return newSnapshot
+}
+
+func (s pgSnapshot) markInProgress(txid uint64) pgSnapshot {
+	if txid >= s.xmax {
+		// Nothing to do
+		return s
+	}
+
+	newSnapshot := pgSnapshot{
+		s.xmin,
+		s.xmax,
+		s.xipList,
+		pgtype.Present,
+	}
+
+	// Adjust the xmax and running tx if necessary
+	if txid < s.xmin {
+		// Adjust the xmin and prepend the newly running tx
+		newSnapshot.xmin = txid
+		newSnapshot.xipList = append([]uint64{txid}, newSnapshot.xipList...)
+	} else {
+		// Add the newly in-progress xip to the list of in-progress transactions
+		pos, found := slices.BinarySearch(newSnapshot.xipList, txid)
+		if !found {
+			newSnapshot.xipList = slices.Insert(newSnapshot.xipList, pos, txid)
+		}
+	}
+
+	// Adjust the xmax if necessary
+	var numToDrop int
+	startingXipLen := len(newSnapshot.xipList)
+	for numToDrop = 0; numToDrop < startingXipLen; numToDrop++ {
+		if newSnapshot.xipList[startingXipLen-1-numToDrop] != newSnapshot.xmax-uint64(numToDrop)-1 {
+			break
+		}
+	}
+
+	if numToDrop > 0 {
+		newSnapshot.xmax = newSnapshot.xipList[startingXipLen-numToDrop]
+		newSnapshot.xipList = newSnapshot.xipList[:startingXipLen-numToDrop]
+		if len(newSnapshot.xipList) == 0 {
+			newSnapshot.xipList = nil
+		}
+	}
+
+	return newSnapshot
+}
+
+func (s pgSnapshot) visible(txid uint64) bool {
```

Comment:
maybe `isTransactionVisible`?

Reply:
Doesn't seem very idiomatic go to me.

Reply:
maybe `txvisible` then? to make it clear at the callsite its checking for a transaction?